Remove mainThreadIndex from active tab main track and use threadIndexes everywhere

canova opened the pull request:
We had a `mainThreadIndex` field for the active tab main track before. We were using this to handle the thread clicking and stuff when we didn't have the thread merging. But since we have thread merging now, we don't need to keep that main thread index anymore. We can directly use the threadIndexes set to compute all of these. This also fixes a clicking issue on a merged thread.

[Example profile currently:](https://profiler.firefox.com/public/p5nfhp8h4xf8qz2hfkhwpcx7qvcsg37rcyr417g/calltree/?profileName=cnn.com%20Nov%202020%2C%20WebDev%20mode&resources&thread=10&v=5&view=active-tab) Try to click on the main thread. It doesn't show the merged thread correctly on the call tree panel.
[Example profile on deploy preview](https://deploy-preview-3066--perf-html.netlify.app/public/p5nfhp8h4xf8qz2hfkhwpcx7qvcsg37rcyr417g/calltree/?profileName=cnn.com%20Nov%202020%2C%20WebDev%20mode&resources&thread=10&v=5&view=active-tab)

gregtatum commented on src/profile-logic/active-tab.js:
```diff
@@ -61,6 +61,8 @@ export function computeActiveTabTracks(
   // Global tracks that are certainly global tracks.
+  // FIXME: We should revert back to full view if we failed to find a track
+  // index for the main track.
   const mainTrackIndexes = [];
```
Question: What happens here if an empty array is used?

canova replied:
If the array is empty, we should abort and revert to full view because there is nothing to show. It's not handled currently, and there is this fixme to remind me :) I have a WIP patch for that as well, probably I'll open a PR for that tomorrow.

gregtatum commented on src/profile-logic/active-tab.js:
```diff
@@ -127,9 +129,6 @@ export function computeActiveTabTracks(
   const mainTrackIndexesSet = new Set(mainTrackIndexes);
```
Suggestion: This is a behavior change. Is it OK for this Set to be empty? To maintain old behavior, if it's empty, then the `0` index should be added. I'm not sure how this is used, but it would be good to ensure this doesn't regress or cause an error somewhere else.

gregtatum commented on src/test/components/ActiveTabTimeline.test.js:
```diff
@@ -93,7 +93,7 @@ describe('ActiveTabTimeline', function() {
         throw new Error('Expected a tab track.');
       }
-      const threadIndex = track.mainThreadIndex;
+      const threadIndex = (track.threadIndexes.values().next().value: any);
```
Suggestion: This `any` coercion is pretty risky, and may introduce an run-time error. This would be better to wrap up in a type-safe function in `src/utils/flow.js` and then properly handle the `void` case.

```
export function getFirstItemFromSet<T>(set: Set<T>): T | void {
  return set.values().next().value;
}
```

gregtatum replied:
Update: Oh, I just realized this is in a test, so the risk is much lower, but it would be good not to coerce to any. So the above suggestion on adding a function is optional. I know we have this awkward `value().next().value` pattern a few times in the codebase, so it might be worth fixing with a good utility function.

gregtatum commented on src/test/components/ActiveTabTimeline.test.js:
```diff
@@ -93,7 +93,7 @@ describe('ActiveTabTimeline', function() {
-      const threadIndex = track.mainThreadIndex;
+      const threadIndex = (track.threadIndexes.values().next().value: any);
 
       if (threadIndex !== null) {
```
Suggestion: I believe this check is now wrong. `threadIndex` may be `undefined`.

gregtatum reviewed (approved):
Thanks for the quick fix!

canova commented:
Thanks Greg for the review! I added a small commit that addresses your feedback about `getFirstItemFromSet`. It changes a few places now, would you mind taking a look at that commit even though you r+'ed already? https://github.com/firefox-devtools/profiler/pull/3066/commits/2c2eff9e7165c798da7db614ea554f25f361f4c7

gregtatum reviewed (approved):
The new commit looks good to me, thanks for it!
